sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/ContainerRegistryRedirectPolicy.java

```diff
@@ -0,0 +1,139 @@
+import com.azure.core.http.HttpRequest;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.policy.HttpPipelinePolicy;
+import com.azure.core.util.CoreUtils;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.publisher.Mono;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.azure.containers.containerregistry.implementation.UtilsImpl.DOCKER_DIGEST_HEADER_NAME;
+
+/**
+ * <p> Redirect policy for the container registry.</p>
+ *
+ * <p> This reads some of the headers that are returned from the redirect call that core redirect policy does not handle.</p>
+ */
+public final class ContainerRegistryRedirectPolicy implements HttpPipelinePolicy {
+    private static final ClientLogger LOGGER = new ClientLogger(com.azure.core.http.policy.DefaultRedirectStrategy.class);
+    private static final int MAX_REDIRECT_ATTEMPTS;
+    private static final String REDIRECT_LOCATION_HEADER_NAME;
+    private static final int PERMANENT_REDIRECT_STATUS_CODE;
+    private static final int TEMPORARY_REDIRECT_STATUS_CODE;
+    private static final Set<HttpMethod> REDIRECT_ALLOWED_METHODS;
+    private static final String AUTHORIZATION;
+
+    static {
+        REDIRECT_ALLOWED_METHODS = new HashSet<>(Arrays.asList(HttpMethod.GET, HttpMethod.HEAD));
+        PERMANENT_REDIRECT_STATUS_CODE = 308;
+        TEMPORARY_REDIRECT_STATUS_CODE = 307;
+        REDIRECT_LOCATION_HEADER_NAME = "Location";
+        MAX_REDIRECT_ATTEMPTS = 3;
+        AUTHORIZATION = "Authorization";
+    }
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        return this.attemptRedirect(context, next, context.getHttpRequest(), 1, new HashSet<>());
+    }
+
+    /**
+     * Function to process through the HTTP Response received in the pipeline
+     * and redirect sending the request with new redirect url.
+     */
+    private Mono<HttpResponse> attemptRedirect(HttpPipelineCallContext context, HttpPipelineNextPolicy next, HttpRequest originalHttpRequest, int redirectAttempt, Set<String> attemptedRedirectUrls) {
+        context.setHttpRequest(originalHttpRequest.copy());
+        return next.clone().process().flatMap((httpResponse) -> {
+            if (this.shouldAttemptRedirect(context, httpResponse, redirectAttempt, attemptedRedirectUrls)) {
+                HttpRequest redirectRequestCopy = this.createRedirectRequest(httpResponse);
+                return httpResponse.getBody().ignoreElements()
+                    .then(this.attemptRedirect(context, next, redirectRequestCopy, redirectAttempt + 1, attemptedRedirectUrls))
+                    .flatMap(newResponse -> {
+                        String digest = httpResponse.getHeaders().getValue(DOCKER_DIGEST_HEADER_NAME);
+                        if (digest != null) {
+                            newResponse.getHeaders().add(DOCKER_DIGEST_HEADER_NAME, digest);
```

Comment:
Should this be `newResponse.getHeaders().set()` or `add()`? `add()` will add another value to the list of values for this header and `set` will overwrite the old value with this new value. Do we want to overwrite or add another value here?